### ISSUE TITLE
Win32 apps look beautiful, Metro apps look horrible

### ISSUE THREAD

**Issue by mufunyo** [snowie2000/mactype#401]
A picture says more than a thousand words... I think something is messing with the gamma on Metro apps. Also, would be nice to be able to force RGB (LCD) mode for Metro apps as well, if possible.

![image](https://user-images.githubusercontent.com/2671260/36225347-ae913cc2-11ca-11e8-9b32-8e65e7e3959a.png)

**Comment by mufunyo** [comment]
MacTypePatch appears to fix this particular issue. It would be nice if you considered integrating the changes in MacTypePatch into the regular build of MacType.

The offending setting that makes text in Metro apps look horrible is called "GrayscaleEnhancedContrast", and if you set GrayscaleEnhancedContrast= 0 in MacTypePatch's UserParams.ini, this is the result:

![image](https://user-images.githubusercontent.com/2671260/36633356-662b8308-1994-11e8-9aa8-ebbeca7893f9.png)

It would still be nice to be able to force RGB mode everywhere, because the grayscale mode that Metro uses still looks worse than native Win32 apps.

**Comment by extratype** [comment]
We need another parameter for GrayscaleEnhancedContrast. It's as same as EnhancedContrast, by the source https://github.com/snowie2000/mactype/blob/f4f1cd3cc27d80f72b411424a3821896d15dff73/directwrite.cpp#L182

**Comment by mufunyo** [comment]
@extratype do you think LCD (ClearType) AA can be forced for all apps including Metro stuff in the code somewhere? For tablets we could implement a rotation callback so that the RGB orientation is adjusted for 90, 180 and 270 degree rotations.

I also tried to find the source code for MacTypePatch 1.26 but the source on GitHub only goes up to version 1.18.

**Comment by extratype** [comment]
@mufunyo I don't think so. The pixel format (RGB vs. black) may not be compatible internally. I'm not sure hooking these [[1]](https://docs.microsoft.com/en-us/windows/desktop/api/d2d1/ne-d2d1-d2d1_text_antialias_mode), [[2]](https://docs.microsoft.com/en-us/windows/desktop/api/dwrite_1/ne-dwrite_1-dwrite_text_antialias_mode) would work.

**Comment by extratype** [comment]
@snowie2000 Thank you for clarification. I just set Contrast=0 and it looks less horrible. But adjusting gamma is not enough to make fonts thicker. Is it possible to override DirectWrite with FreeType so that I can adjust RenderWeight? @silight-jp had some [experiments](https://silight.hatenablog.jp/entry/2016/10/03/234822) regarding that.

**Comment by mufunyo** [comment]
@extratype Wow that is an interesting rabbit hole to dive into. It looks like Microsoft engineers are completely oblivious to how subpixel alpha blending can be done on the GPU, and instead the font engine is asked to render opaque text over a static background, and all ClearType functionality is disabled when the destination surface contains an alpha channel.

Of course subpixel alpha blending is perfectly possible, but you need an RGB alpha buffer rather than monochromatic alpha. It looks like the folks over at [freetype-gl](https://github.com/rougier/freetype-gl/issues/138) have successfully implemented that in OpenGL using a two-pass pixel shader.

Maybe we can hook Direct2D surface rendering to render text on translucent surfaces by keeping track of an RGB alpha buffer and blending it using the two-pass pixel shader? I'm thinking whenever a Direct2D surface is rendered to using a hooked MacType function, a corresponding RGB alpha surface is created for it, if it does not already exist. Then in the hooked Present() or whatever D2D uses, any surface that has been "touched" by MacType in this way is rendered with the two-pass shader, and every other surface is just rendered as it always would be.

**Comment by mufunyo** [comment]
Updating to the latest beta of mainline MacType coming from MacTypePatch has DirectWrite working in Chrome but UWP/Settings don't seem to be affected by MacType at all any more. How much of MacTypePatch has been merged into mainline MacType at this point?

**Comment by extratype** [comment]
I think MacTypePatch has already been merged into MacType 2017.628.0. But, you know, `FontSubstitutes` is disabled ([#310](https://github.com/snowie2000/mactype/issues/310)) and some are linked to other parameters in MacType. You would have to modify source and build your own version to take control all of them, as I did in [my forked repo](https://github.com/extratype/mactype).

**Comment by extratype** [comment]
I leave comment here as issues have been merged into this one.

I tried CentBrowser, DirectWrite could be turned off but still parameters was not applied as Chrome. So I just went back to Chrome. I digged Chromium repository a bit to see what's happening.

Chromium uses [Skia](https://skia.org/) to render text across platforms. The rendering part for Windows is `SkScalerContext_win_dw.cpp`. It calls `IDWriteGlyphRunAnalysis::CreateAlphaTexture` and then performs its own blending. Chromium tries to match with the default DirectWrite rendering, but it is slightly lighter. This problem is discussed [here](https://bugs.chromium.org/p/chromium/issues/detail?id=534732). It ignores most of rendering parameters changed by MacType because Skia does not query `IDWriteGlyphRunAnalysis::GetAlphaBlendParams`. 

The Chromium issue I mentioned above suggests changing `SK_GAMMA_CONTRAST` (it's constant, it can't be changed at runtime). Maybe CentBrowser and Advanced Chrome changed this value to make text thicker? Anyway, there's almost nothing MacType could do to handle this.

For now MacType can be used to force `HintingMode` and `RenderingMode`. I turned off hinting and set `RenderingMode` to `5`. This makes text even more thinner. For workaround you can use an extension to set your own CSS. See #402 for example. I am quite satisfied with this:
```css
body {
    -webkit-text-stroke-width: 0.24px;
    text-shadow: 0 0 rgba(128,128,128,0.2);
    text-rendering: geometricPrecision;
}
```
This makes text look comparable to GDI rendering in MacType for me. It's still rendered in DirectWrite, slightly thicker and brighter, but it looks smooth 😄

**Comment by sammilucia** [comment]
Thanks @extratype ... Just FYI CentBrowser does use GDI for UI and the chrome if you follow the steps in the beta4 release notes https://github.com/snowie2000/mactype/releases/tag/v1.2018.10.19-beta4

**Comment by mufunyo** [comment]
For reference, the screenshot in my 24 feb comment were created with the following settings:

`[General]
HintingMode=1
AntiAliasMode=4
NormalWeight=0
BoldWeight=0
ItalicSlant=0
EnableKerning=1
GammaMode=0
LcdFilter=2
BolderMode=2
TextTuning=0
TextTuningR=0
TextTuningG=0
TextTuningB=0
GammaValue=1.4
Contrast=1.0
RenderWeight=1.0
Fontlink=2
HookChildProcesses=1
FontLoader=0
FontSubstitutes=1
Shadow=1,1,0,0x0,0,0x0
MaxBitmap=0
DirectWrite=1
HintSmallFont=1
[Individual]
Courier New=,,22,,,`

And the following UserParams (MacTypePatch):
`[General]
HookTarget = 0
ForceNoHinting = 0
[Direct2D]
Gamma = 1800
EnhancedContrast = 0
ClearTypeLevel = 100
PixelGeometry = 1
RenderingMode = 5
GrayscaleEnhancedContrast = 0
GridFitMode = 1
RenderingMode1 = 5
[DirectWrite]
Gamma = 1800
EnhancedContrast = 0
ClearTypeLevel = 100
PixelGeometry = 1
RenderingMode = 5
GrayscaleEnhancedContrast = 0
GridFitMode = 1
RenderingMode1 = 5`

I understand that font rendering preferences are subject to personal taste, but at least as far as fonts' internal metrics, as well as a technical standpoint of mathematically & visually correct gamma, this is the most "true to life" rendering of fonts, assuming an sRGB or Rec709 display. Any thickening or thinning of fonts beyond this would be "incorrect", although as mentioned before, personal taste can vary.

**Comment by sammilucia** [comment]
@extratype i really like these settings!
I haven't seen these [DirectWrite] or [Direct2D] settings before and they're different from those in the Wiki ... are these the new ones?

**Comment by extratype** [comment]
@sammilucia those settings are for MacTypePatch.

**Comment by extratype** [comment]
@sammilucia Those settings are for MacTypePatch. They are translated to...

```ini
[General]
HintingMode=1
AntiAliasMode=4

[DirectWrite]
GammaValue=1.8
Contrast=0
ClearTypeLevel=1.0
RenderingMode=5
```

But `ForceNoHinting` conflicts with MacType. Hinting is disabled so it is fixed to 1.

**Comment by sammilucia** [comment]
@extratype right! thank you.... so far i've avoided using DirectWrite! 😜

**Comment by sammilucia** [comment]
@extratype can I please add your profile to the official MacType distribution? 😊 😊

**Comment by extratype** [comment]
@sammilucia Sure! 😄

**Comment by extratype** [comment]
@sammilucia Oops I didn't write it. Please recheck the comments.

**Comment by sammilucia** [comment]
@extratype I have, I can't see who wrote it .. please tell me?

**Comment by mufunyo** [comment]
@silight-jp wrote MacTypePatch, if that's what you mean.

**Comment by extratype** [comment]
@sammilucia @mufunyo wrote https://github.com/snowie2000/mactype/issues/401#issuecomment-431594262.

Here's my profile:
```ini
[General]
HintingMode=1
AntiAliasMode=4
NormalWeight=6
BoldWeight=3
ItalicSlant=0
Saturation=0
UseMapping=0
GammaMode=0
GammaValue=1.4
Contrast=1.1
RenderWeight=1.2
TextTuning=0
TextTuningR=1
TextTuningG=1
TextTuningB=1
BolderMode=0
FontLoader=0
FontLink=1
FontSubstitutes=2
LcdFilter=2
EnableKerning=1
HintSmallFont=0
Shadow=0,0,0,0x0,0,0x0
LoadOnDemand=1
CacheMaxFaces=32
CacheMaxSizes=768
CacheMaxBytes=50331648
MaxBitmap=0
DirectWrite=1

[DirectWrite]
RenderingMode=5
GammaValue=1.4
Contrast=0.625
ClearTypeLevel=0.75
```

**Comment by sammilucia** [comment]
Oh! I'm so sorry @mufunyo!

@mufunyo would you mind if I include your profile in the main release? It's quite unlike any of the other profiles so I think it's a good option for people.

Thanks @extratype I'll try yours too! I will say I'm looking for profiles which are substantially different from the other ones bundled with MacType, so I'll give it a try.

Your cache settings look unusual @extratype, you're a dev, so I assume there's a good reason for them?

**Comment by mufunyo** [comment]
Ohh, we're talking about who wrote the profile, not the patch, haha. I'm sorry, I only quickly skimmed the comments today and missed what this was about.

Yeah, I wrote that profile. Feel free to include it! I think the gamma difference between FreeType (1.8) and DirectWrite (1.4) is probably because of a bug/quirk in DirectWrite (not unusual for Microsoft), as they visually match each other. 2.2 would be the closest pure power-law gamma value to sRGB/Rec709, however it is immediately obvious this is an incorrect value for both FreeType and DirectWrite, so manual (visual) tuning was used to reach the final values of 1.8 and 1.4 respectively. It might be because near black sRGB gamma slopes towards 1.0, but I'm really not well versed enough in the maths behind it to be sure.

**Comment by sammilucia** [comment]
@mufunyo thank you! It will be included in the next beta!

I compared your DW and FT gamma and agreed with it visually. It will be the non MacType-Patch version that's included. We've asked @slight-jp to join the team a couple of times but can't seem to get a response :disappointed: 

Yes, Microsoft don't seem to care about how fonts look... 😜

**Comment by extratype** [comment]
@sammilucia I wanted to make fonts smooth and more easier to read like Mac. I chose Source Han Sans (converted to ttf) as an alternative system default font. I tried built-in profiles but they looked too thick and some parameters felt like extreme to me. I googled for some good articles and heuristically tweaked parameters like them to meet my taste and make fonts similar to Mac. I was not rigorous and actually it ended a bit thinner than Mac but I felt comfortable so I am sticking with it.

Regarding cache settings, Source Han Sans was heavy so I just adjusted them bigger, hoping for lower latency.

**Comment by mufunyo** [comment]
As an aside, I tested El Capitan a while back in a VM and none of the MacType profiles came even close to matching what Apple does with small type fonts today. I can't even say I really like it, because there is very strong grid fitting, but in a radically different way than how Microsoft or FreeType do it. It seems to be either blending a grid fitted version with a non grid fitted version, or dilating the font weight after grid fitting. Curiously this behaviour is disabled when LCD anti-aliasing is disabled.

For reference, here is the same text twice, once with LCD anti-aliasing enabled (and converted to greyscale for comparison) and once with it disabled. With it disabled, the appearance matches what you would expect for a non grid fitted rendering of small type. With it enabled however, there is strong grid fitting (note the capital C in Click) but there are still fractional pixels outside the grid. By default, the gamma setting is also much darker.

![image](https://user-images.githubusercontent.com/2671260/48212425-d93df900-e37b-11e8-83ee-5cbca296e81a.png)

![image](https://user-images.githubusercontent.com/2671260/48212438-e22eca80-e37b-11e8-85c5-a1fef0570861.png)

Apparently this behaviour started with Yosemite, and versions before that match the way FreeType renders with grid fitting disabled.

**Comment by sammilucia** [comment]
@extratype to me your DirectWrite settings look quite different from GDI ... but then viewing English not Chinese ... ?

**Comment by sammilucia** [comment]
@mufunyo yes I'm not a fan of tight grid alignment either. It ruins the integrity of the font. Would never have happened if Steve was still alive! 😂

😭

**Comment by extratype** [comment]
@sammilucia I can't match it with GDI while keeping letters clear. You can't make them thicker without ruining the anti-aliasing (DirectWrite sucks again). So I took a compromise.

**Comment by sammilucia** [comment]
Right. So maybe we could describe this profile as Soft and clear (or something) for Chinese text?

**Comment by extratype** [comment]
@sammilucia I didn't care about languages when I wrote my profile. I speak Korean though.

**Comment by mufunyo** [comment]
@snowie2000 Personally I think it accomplishes neither; it doesn't follow the original shapes of the font because it's grid fitted, and the dilation or blending (whichever method is actually used) makes the font fuzzy regardless. It seems like the kind of compromise-by-committee that produces the ugly cars you see on the road today. Like @sammilucia said, wouldn't have happened if Steve was still alive.

Greyscale rendering also doesn't have anything to do with it other than the fact that Apple seems to fall back to old method of non grid fitted rendering in greyscale mode. If you look at Mavericks, aside from the different system font (Lucida Grande was later replaced by a new font that supposedly looks better on Retina displays) there is a combination of LCD subpixel rendering and non grid fitted rendering, which is what OS X prior to Yosemite has always used:

![image](https://user-images.githubusercontent.com/2671260/48263523-ac92eb80-e426-11e8-9944-2d386e0a334d.png)

**Comment by snowie2000** [comment]
This is a setting  in GDIPlus  helium version it  doesn't work in mactype

**Comment by extratype** [comment]
I set `RenderWeight=1.25` to match with printed text just by looking.

Edit: Also set
```ini
[DirectWrite]
RenderingMode=5
GammaValue=1.3
Contrast=0.6
ClearTypeLevel=0.7
```

**Comment by mufunyo** [comment]
That reminds me, I recently finally installed 2018.1-beta5 because I was reinstalling Windows anyway, and I noticed the "hint small fonts" option is either broken, or the threshold for when fonts get hinted is set to a smaller font size than previously. I've gotten used to it by now, but it does hurt legibility in apps like Explorer slightly.
